src/main/java/net/fabricmc/loom/task/MigrateMappingsTask.java

```diff
@@ -61,123 +68,130 @@ public void doTask() throws Throwable {
+		} catch (IOException e) {
+			throw new IllegalArgumentException("Could not find mappings for version " + targetMappingsVersion, e);
 		}
+	}
 
+	private static void migrateMappings(Project project, MinecraftMappedProvider minecraftMappedProvider,
+										Path inputDir, Path outputDir, TinyTree currentMappings, TinyTree targetMappings, LoomGradleExtension extension
+	) throws IOException {
 		project.getLogger().lifecycle(":joining mappings");
-		MappingSet mappingSet = new MappingsJoiner(sourceMappings, targetMappings, "intermediary", "named").read();
+		MappingSet mappingSet = new MappingsJoiner(currentMappings, targetMappings,
+						"intermediary", "named").read();
 
 		project.getLogger().lifecycle(":remapping");
-		Mercury mercury = new Mercury();
+		Mercury mercury = SourceRemapper.createMercuryWithClassPath(project, false);
 
-		for (File file : project.getConfigurations().getByName(Constants.MINECRAFT_DEPENDENCIES).getFiles()) {
-			mercury.getClassPath().add(file.toPath());
-		}
-
-		for (File file : project.getConfigurations().getByName("compileClasspath").getFiles()) {
-			mercury.getClassPath().add(file.toPath());
-		}
+		mercury.getClassPath().add(minecraftMappedProvider.MINECRAFT_MAPPED_JAR.toPath());
+		mercury.getClassPath().add(minecraftMappedProvider.MINECRAFT_INTERMEDIARY_JAR.toPath());
 
 		mercury.getClassPath().add(extension.getMinecraftMappedProvider().MINECRAFT_MAPPED_JAR.toPath());
 		mercury.getClassPath().add(extension.getMinecraftMappedProvider().MINECRAFT_INTERMEDIARY_JAR.toPath());
 
-		mercury.getProcessors().add(MercuryRemapper.create(mappingSet));
-
 		try {
-			mercury.rewrite(inputDir.toPath(), outputDir.toPath());
+			mercury.rewrite(inputDir, outputDir);
 		} catch (Exception e) {
 			project.getLogger().warn("Could not remap fully!", e);
 		}
 
-		project.getLogger().lifecycle(":cleaning file descriptors");
-		System.gc();
+		try {
+			mercury.rewrite(inputDir, outputDir);
+		} catch (Exception e) {
+			project.getLogger().warn("Could not remap fully!", e);
+		}
 	}
 
 	public static class MappingsJoiner extends MappingsReader {
-		private final Mappings sourceMappings, targetMappings;
+		private final TinyTree sourceMappings, targetMappings;
 		private final String fromNamespace, toNamespace;
 
-		public MappingsJoiner(Mappings sourceMappings, Mappings targetMappings, String fromNamespace, String toNamespace) {
+		/**
+		 * Say A is the source mappings and B is the target mappings.
+		 * It does not map from intermediary to named but rather maps from named-A to named-B, by matching intermediary names.
+		 * It goes through all of the intermediary names of A, and for every such intermediary name, call it I,
+		 * matches the named mapping of I in A, with the named mapping of I in B.
+		 * As you might imagine, this requires intermediary mappings to be static across all versions, which they are. :volderthonk:
```

Comment:
Remove `:volderthonk:` no need for it here, its fine on discord.

Reply:
This is a great travesty